docs/dev/design/new-notifications-system.rst

```diff
@@ -0,0 +1,459 @@
+                slug="subscription-update-your-cc-details",
+                state__in=[UNREAD, READ],
+                attached_to=Organization,
+                attached_to_id=organization.id,
+            ).update(state=DISMISSED)
+
+
+
+API definition
+--------------
+
+I will follows the same pattern we have on APIv3 that uses nested endpoints.
+This means that we will add a ``/notifications/`` postfix to most of the resource endpoints
+where we want to be able to attach/list notifications.
+
+Notifications list
+~~~~~~~~~~~~~~~~~~
+
+.. http:get:: /api/v3/users/(str:user_username)/notifications/
+
+    Retrieve a list of all the notifications for this user.
+
+.. http:get:: /api/v3/projects/(str:project_slug)/notifications/
+
+    Retrieve a list of all the notifications for this project.
+
+.. http:get:: /api/v3/organizations/(str:organization_slug)/notifications/
+
+    Retrieve a list of all the notifications for this organization.
+
+.. http:get:: /api/v3/projects/(str:project_slug)/builds/(int:build_id)/notifications/
+
+    Retrieve a list of all the notifications for this build.
+
+    **Example response**:
+
+    .. sourcecode:: json
+
+        {
+            "count": 25,
+            "next": "/api/v3/projects/pip/builds/12345/notifications/?unread=true&sort=type&limit=10&offset=10",
+            "previous": null,
+            "results": [
+                {
+                    "slug": "cancelled-by-user",
+                    "state": "unread",
+                    "dismissable": false,
+                    "bell": false,
+                    "attached_to": "build",
+                    "message": {
+                        "header": "User action",
+                        "body": "Build cancelled by the user.",
+                        "type": "error",
+                        "icon": "fa-exclamation",
+                        "color": "red"
+                    }
+                }
+            ]
+        }
+
+    :query boolean unread: return only unread notifications
+    :query string type: filter notifications by type (``error``, ``note``, ``tip``)
+    :query string sort: sort the notifications (``type``, ``date`` (default))
+
+
+Notification create
+~~~~~~~~~~~~~~~~~~~
+
+
+.. http:post:: /api/v3/projects/(str:project_slug)/builds/(int:build_id)/notifications/
+
+    Create a notification for the resource.
+    In this example, for a ``Build`` resource.
+
+    **Example request**:
+
+    .. sourcecode:: json
+
+        {
+            "slug": "cancelled-by-user",
+            "type": "error",
+            "state": "unread",
+            "dismissable": false,
+            "bell": false,
+        }
+
+
+.. note::
+
+   Similar API endpoints will be created for each of the resources
+   we want to attach a ``Notification`` (e.g. ``User``, ``Organization``, etc)
+
+
+Backward compatibility
+----------------------
+
+It's not strickly required, but if we want, we could extract the current notification logic from:
```

Comment:
I'd say this is probably a requirement.

In fact, I might say these should be the first notifications to more to the new system, as I've deleted this logic from the new templates entirely already.

With notification logic in the application and in templates, adding a third layer on top will be a bit overwhelming. We can completely remove the template "notification" logic, and relatively easily too.

Reply:
The other point of this "backward compatibility" is about migrating the "old build notifications" to the new `Notification` model we are defining here. I wouldn't start here either as first step without doing some testing on production.

Old builds will keep showing old notifications based on `Build.error` fields. New builds will show only notifications from the new system.

Reply:
I'm not sure if I'm being clear here or not, but I don't plan to touch the old templates at all 😄 

When I said "extract the current notification logic" I mean "re-implement the logic that lives in the template in the backend and create a `Notification`" --but I'm not going to touch the template or remove that logic from them.

Reply:
Yeah, that all makes sense and matches my understanding of what we want here too. I think we can pretty easily avoid doing anything that will affect legacy builds, though the new dashboard JS or templates will need to accommodate both patterns obviously. This was expected at least though.